shotover/src/codec/cassandra.rs

```diff
@@ -67,96 +153,246 @@ impl CassandraDecoder {
+                        unimplemented!("Cannot support non-self contained frames yet");
+                    }
+
+                    frame_bytes.advance(UNCOMPRESSED_FRAME_HEADER_LENGTH);
+                    let mut payload = frame_bytes.split_to(payload_length);
+
+                    let mut envelopes: Vec<Message> = vec![];
+
+                    while !payload.is_empty() {
+                        let body_len =
+                            i32::from_be_bytes(payload[5..9].try_into().unwrap()) as usize;
+
+                        let envelope_len = ENVELOPE_HEADER_LEN + body_len;
+
+                        if envelope_len > payload.len() {
+                            return Err(anyhow!(format!(
+                                "envelope length {} is longer than payload length {}",
+                                envelope_len,
+                                payload.len()
+                            ),));
+                        }
+
+                        let envelope = payload.split_to(envelope_len);
+
+                        envelopes.push(Message::from_bytes(
+                            envelope.freeze(),
+                            crate::message::ProtocolType::Cassandra {
+                                compression: Compression::None,
+                            },
+                        ));
+                    }
+
+                    Ok(envelopes)
+                }
+                _ => unimplemented!(),
+            },
+            (_, _) => {
+                let bytes = src.split_to(frame_len);
+                tracing::debug!(
+                    "{}: incoming cassandra message:\n{}",
+                    self.direction,
+                    pretty_hex::pretty_hex(&bytes)
+                );
+
+                let compressed = self.check_compression(&bytes).unwrap();
+
+                let message = Message::from_bytes(
+                    bytes.freeze(),
+                    crate::message::ProtocolType::Cassandra {
+                        compression: if compressed {
+                            compression
+                        } else {
+                            Compression::None
+                        },
+                    },
+                );
+
+                Ok(vec![message])
+            }
+        }
+    }
+
+    fn check_size(
+        &self,
+        src: &BytesMut,
+        version: Version,
+        compression: Compression,
+        handshake_complete: bool,
+    ) -> Result<usize, CheckFrameSizeError> {
+        match (version, handshake_complete) {
+            (Version::V5, true) => match compression {
+                Compression::None => {
+                    let buffer_len = src.len();
+                    if buffer_len < UNCOMPRESSED_FRAME_HEADER_LENGTH {
+                        return Err(CheckFrameSizeError::NotEnoughBytes);
+                    }
+
+                    let header = i64::from_le_bytes(src[..8].try_into().unwrap()) & 0xffffffffffff; // convert to 6 byte int
```

Comment:
The buffer_len check uses UNCOMPRESSED_FRAME_HEADER_LENGTH which is 6
However here we are exceeding that by reading 8 bytes.

I think the best fix here would be:
```rust
let payload_length = (u32::from_le_bytes(src[..4].try_into().unwrap()) & 0x1ffff) as usize;
```

Reply:
Yeah I'm fine with that, and I assume you mean use that in both `check_size` and `decode_frame`?